src/cdomains/unionFind.ml

```diff
@@ -10,10 +10,35 @@ module M = Messages
 
 exception Unsat
 
-type ('v, 't) term = Addr of 'v | Aux of 'v * 't | Deref of ('v, 't) term * Z.t * 't [@@deriving eq, ord, hash]
-type ('v, 't) prop = Equal of ('v, 't) term * ('v, 't) term * Z.t | Nequal of ('v, 't) term * ('v, 't) term * Z.t
-                   | BlNequal of ('v, 't) term * ('v, 't) term
-[@@deriving eq, ord, hash]
+(* equality of terms should not depend on the expression *)
+let compare_exp _ _ = 0
+let equal_exp _ _ = true
+let hash_exp _ = 1
+
+type term = Addr of Var.t | Aux of Var.t * exp | Deref of term * Z.t * exp [@@deriving eq, hash, ord]
+
+let normal_form_tuple_3 (t1,t2,z) =
+  let cmp = compare_term t1 t2 in
+  if cmp < 0 || (cmp = 0 && Z.geq z Z.zero) then (t1,t2,z) else
+    (t2,t1,Z.(-z))
+
+let normal_form_tuple_2 (t1,t2) =
+  if compare_term t1 t2 < 0 then (t1,t2) else
+    (t2,t1)
+
+(** Two propositions are equal if they are syntactically equal
+    or if one is t_1 = z + t_2 and the other t_2 = - z + t_1. *)
+let tuple3_equal p1 p2 = Tuple3.eq equal_term equal_term Z.equal (normal_form_tuple_3 p1) (normal_form_tuple_3 p2)
+let tuple3_cmp p1 p2 = Tuple3.comp compare_term compare_term Z.compare (normal_form_tuple_3 p1) (normal_form_tuple_3 p2)
+let tuple3_hash p1 = Hashtbl.hash (normal_form_tuple_3 p1)
+let tuple2_equal p1 p2 = Tuple2.eq equal_term equal_term (normal_form_tuple_2 p1) (normal_form_tuple_2 p2)
+let tuple2_cmp p1 p2 = Tuple2.comp compare_term compare_term (normal_form_tuple_2 p1) (normal_form_tuple_2 p2)
+let tuple2_hash p1 = Hashtbl.hash (normal_form_tuple_2 p1)
+
+type prop = Equal of ((term * term * Z.t) [@equal tuple3_equal][@compare tuple3_cmp][@hash tuple3_hash])
+          | Nequal of ((term * term * Z.t) [@equal tuple3_equal][@compare tuple3_cmp][@hash tuple3_hash])
+          | BlNequal of ((term * term) [@equal tuple2_equal][@compare tuple2_cmp][@hash tuple2_hash])
```

Comment:
It might be worth giving (semantic) names to the types `term * term * Z.t` and `term * term` above and use them here. That would avoid all the extra attributes here.